sql,pgwire: fix timestamp and tz offset format

rafiss opened the pull request:
This change includes the following:
- Make data type size for time[stamp][tz] match Postgres in pg_catalog and over pgwire.
- Change parameter status sending so it occurs after sending CommandComplete.
- Normalize "UTC" and "GMT" to uppercase when storing timezone offset.

The new tests cover all of these.

Epic: None
Release note (bug fix): Adjusted the size reported for TIME, TIMETZ, TIMESTAMP, TIMESTAMPTZ, and DATE types in the pgwire protocol.

rafiss committed:
sql,pgwire: fix timestamp and tz offset format

This change includes the following:
- Make data type size for time[stamp][tz] match Postgres in pg_catalog
  and over pgwire.
- Change parameter status sending so it occurs after sending
  CommandComplete.
- Normalize "UTC" and "GMT" to uppercase when storing timezone offset.

The new tests cover all of these.

Release note (bug fix): Adjusted the size reported for TIME, TIMETZ, TIMESTAMP,
TIMESTAMPTZ, and DATE types in the pgwire protocol.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97145)
<!-- Reviewable:end -->

otan commented on pkg/sql/pgwire/testdata/pgtest/timezone:
```diff
@@ -5,8 +5,8 @@ Query {"String": "SET TIME ZONE \"UTC\""}
 ----
-{"Type":"ParameterStatus","Name":"TimeZone","Value":"UTC"}
 {"Type":"CommandComplete","CommandTag":"SET"}
+{"Type":"ParameterStatus","Name":"TimeZone","Value":"UTC"}
```
i have just discovered PG doesn't send this out if the value of the timezone didn't change, but that's not a now problem...

otan commented on pkg/sql/pgwire/testdata/pgtest/timezone:
```diff
@@ -52,8 +52,8 @@ Query {"String": "SET TIME ZONE \"America/Chicago\""}
 ----
-{"Type":"ParameterStatus","Name":"TimeZone","Value":"America/Chicago"}
 {"Type":"CommandComplete","CommandTag":"SET"}
+{"Type":"ParameterStatus","Name":"TimeZone","Value":"America/Chicago"}
```
i feel like some other test files need to change as a result of this change. how did they pass `pgwire` datadriven tests before o_o

rafiss replied:
this file wasn't passing against PG before 🤥 

let's see which others turn up

otan reviewed (approved):
i think some other test files with ParamStatusUpdates will need to change but otherwise LGTM

rafiss committed:
sql: only send paramater status update if the value changed

Release note (bug fix): The ParameterStatus message is now only sent
over the pgwire protocol if the value of the parameter changed. (The
parameters that are sent this way are timezone, intervalstyle,
datestyle, is_superuser, and application_name.)

otan commented on pkg/sql/exec_util.go:
```diff
@@ -2929,11 +2929,14 @@ func (m *sessionDataMutator) bufferParamStatusUpdate(param string, status string
 	}
-	m.bufferParamStatusUpdate("application_name", appName)
+	if oldName != appName {
+		m.bufferParamStatusUpdate("application_name", appName)
```
i wonder if we should make `bufferParamStatusUpdate` always take in the old value instead and do it all in one place.

otan replied:
hmm nvm, for `is_superuser` we always have to send that over.

rafiss commented:
tftr!

bors r=otan
